mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+            normalized_tokens (List[Dict]): Normalized tokens represented as dictionaries.
+                For Example:
+                    norm_token = {
+                        "entity": "order",
+                        "raw_entity": "order",
+                        "raw_token_index": 1,
+                        "raw_start": 1
+                    }`
+        """
+        if not self.normalizers:
+            logger.warning(
+                "This instance of TextPreparationPipeline does not have normalizers."
+                "Returning raw tokens."
+            )
+            return raw_tokens
+
+        normalized_tokens = []
+        for i, raw_token in enumerate(raw_tokens):
+            if not raw_token["text"]:
+                continue
+            normalized_text = self._normalize_single_token(raw_token["text"])
+            if len(normalized_text) > 0:
+                for token in normalized_text.split():
+                    norm_token = {}
+                    norm_token["entity"] = token
+                    norm_token["raw_entity"] = raw_token["text"]
+                    norm_token["raw_token_index"] = i
+                    norm_token["raw_start"] = raw_token["start"]
+                    normalized_tokens.append(norm_token)
+        return normalized_tokens
+
+    def _normalize_single_token(self, text):
+        """Normalize an individual token by processing text with all normalizers.
+
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_text (str): Normalized text.
+        """
+        normalized_text = text
+        for normalizer in self.normalizers:
+            normalized_text = (
+                TextPreparationPipeline.modify_around_mindmeld_annotations(
+                    text=normalized_text, function=normalizer.normalize
+                )
+            )
+        return normalized_text
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return TextPreparationPipeline.tokenize_around_mindmeld_annotations(
+            text=text, function=self.tokenizer.tokenize
+        )
+
+    def stem_word(self, word):
+        """
+        Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.
+
+        Args:
+            words (List[str]): List of words to stem.
+
+        Returns:
+            stemmed_words (List[str]): List of stemmed words.
+        """
+        return self.stemmer.stem_word(word)
+
+    @staticmethod
+    def find_mindmeld_annotation_re_matches(text):
+        """
+        Args:
+            text (str): The string to find mindmeld annotation instances
+                (" {entity_text|entity_type} ")
+        Returns:
+            matches (List[sre.SRE_Match object]): Regex match objects.
+        """
+        COMPILED_MINDMELD_ANNOTATION_PATTERN = re.compile(MINDMELD_ANNOTATION_PATTERN)
```

Comment:
I think we should just compile it right after assigning `MINDMELD_ANNOTATION_PATTERN `

Reply:
Makes sense! Now `MINDMELD_ANNOTATION_PATTERN` represents the compiled form